torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp

```diff
@@ -0,0 +1,370 @@
+  auto seg_begin = exprs.begin();
+  auto seg_end = exprs.begin();
+  Score seg_score = scores.at(*seg_begin);
+  while (seg_end != exprs.end()) {
+    const auto expr = *seg_end;
+    const auto cur_score = scores.at(expr);
+    if (seg_score == cur_score) {
+      // advance further
+      ++seg_end;
+      continue;
+    } else if (seg_score < cur_score) {
+      // segment ended
+      reorderSegmentBreadthFirst(seg_begin, seg_end);
+      seg_begin = seg_end;
+      seg_score = cur_score;
+    } else {
+      // exprs list is assumed to be sorted in the order of scores, so
+      // this should never be reachable
+      TORCH_INTERNAL_ASSERT(
+          false, "Unexpected expression: ", expr, ", score: ", cur_score);
+    }
+  }
+  reorderSegmentBreadthFirst(seg_begin, seg_end);
+}
+
+void mergeNonRootGroupsIntoRootGroups(
+    TargetGroupMap& computed_at_exprs,
+    ExprTargetMap& target_map) {
+  for (auto it = computed_at_exprs.begin(); it != computed_at_exprs.end();) {
+    TensorView* target = it->first;
+    if (target->hasComputeAt()) {
+      Expr* target_expr = target->definition();
+      TensorView* target_of_target = target_map.at(target_expr);
+      auto& target_group = computed_at_exprs.at(target_of_target);
+      auto pos =
+          std::find(target_group.begin(), target_group.end(), target_expr);
+      TORCH_INTERNAL_ASSERT(pos != target_group.end());
+      target_group.insert(pos, it->second.begin(), it->second.end());
+      // Update the target map
+      for (auto& inserted_expr : it->second) {
+        TORCH_INTERNAL_ASSERT(target_map.at(inserted_expr) == target);
+        target_map.at(inserted_expr) = target_of_target;
+      }
+      it = computed_at_exprs.erase(it);
+    } else {
+      ++it;
+    }
+  }
+}
+
+// Merge root loop-nests into reordered_exprs
+void mergeGroupsIntoSortedList(
+    TargetGroupMap& computed_at_exprs,
+    ExprList& reordered_exprs) {
+  while (computed_at_exprs.size() > 0) {
+    // Find the root loop-nest that has no dependency with the other
+    // loop-nests
+    TensorView* cur_target = computed_at_exprs.begin()->first;
+    for (auto& group : computed_at_exprs) {
+      auto target = group.first;
+      if (cur_target == target)
+        continue;
+      if (DependencyCheck::isDependencyOf(target, cur_target)) {
+        cur_target = target;
+      }
+    }
+    // cur_target can be visited
+    reordered_exprs.insert(
+        reordered_exprs.end(),
+        computed_at_exprs.at(cur_target).begin(),
+        computed_at_exprs.at(cur_target).end());
+    computed_at_exprs.erase(cur_target);
+  }
+}
+
+} // namespace
+
+// Reorder exprs so that LoopNestGenerator::handle(Expr*) can generate
+// correct loop nests. Vector exprs is assumed to be topologically
+// sorted, but that is not sufficient as tensors computed at
+// outer loops need to be located earlier.
+std::vector<Expr*> reorderExprsForComputeAt(const std::vector<Expr*>& exprs) {
```

Comment:
if this is considered a pass it definitely needs a perf scope

Reply:
Added